### ISSUE TITLE
Weird scipy warning in t-test when there are all-zero genes

### ISSUE THREAD

**Issue by gokceneraslan** [scverse/scanpy#629]
After we switch to scipy t-test, I started getting a scipy warning (`invalid value encountered`) when I run the following the code:

```python
adata = sc.datasets.paul15()
adata.X[:, 10] = 0.0
sc.tl.rank_genes_groups(adata, 'paul15_clusters')
```

Output:

![image](https://user-images.githubusercontent.com/1140359/57115885-88ea9700-6d1f-11e9-8752-a865cda602b7.png)

@ivirshup is having a look now.

**Comment by ivirshup** [comment]
It looks like those warning are being raised from `scipy.stats.distributions.t.sf`. 

This was also happening in the tests, but there's already a bunch of warnings in the tests so we didn't see it.

~~I believe we didn't get this warning from the older code because of these lines:~~

```python
dof[np.isnan(dof)] = 0		
pvals = stats.t.sf(abs(scores), dof)*2 # *2 because of two-tailed t-test
```

I don't think it's the above lines anymore, since the replacing the `ttest_ind_from_stats` call with the following still throws the warning:

```python
df, denom = stats.stats._unequal_var_ttest_denom(
    v1=var_group, n1=ns_group, v2=var_rest, n2=ns_rest
)
df[np.isnan(df)] = 0
scores, pvals = stats.stats._ttest_ind_from_stats(
    mean_group, mean_rest, denom, df
)
```

Other than that, potential solutions include:

*  Mask out genes which aren't expressed in the compared groups (since there's not too much point in getting and correcting a pvalue for them)
* Revert change (would bring back issue of genes with variance of 0)
* Wrap the t-test with something like `np.errstate` to hide the warning

**Comment by gokceneraslan** [comment]
Cool! 

> * Mask out genes which aren't expressed in the compared groups (since there's not too much point in getting and correcting a pvalue for them)

I think masking out might be problematic because, `n_genes=adata.n_vars` should return all genes in any case. 

> * Revert change (would bring back issue of genes with variance of 0)

I feel like using scipy function will slightly increase the maintainability (and simplicity) of the code, so I'm fine with keeping the scipy switch.

> * Wrap the t-test with something like `np.errstate` to hide the warning

This sounds good. Replacing weird scipy warning with a proper scanpy warning would also make sense.

**Comment by LuckyMD** [comment]
You could just output `NaN` for all genes that were masked? That would be accurate as the test would fail in that case anyway. That should solve `n_genes == adata.n_vars`

**Comment by ivirshup** [comment]
Definitely agree that the shape should be the same. Would it be unreasonable to give all zero genes a p-value of one, since they're definitely not differentially expressed?

**Comment by ivirshup** [comment]
Closed via #683
